src/System.Text.Primitives/System/Text/InvariantParser_int.cs

```diff
@@ -31,68 +32,139 @@ public static bool TryParse(byte[] utf8Text, int index, out int value, out int b
+                    else if (value > 0 && (Int32.MaxValue - value * 10) + (negative ? 1 : 0) < nextByteVal) // overflow (+1 allows for overflow for min value)
                     {
-                        value = default(int); // if we haven't, set value to 0 and return false
+                        value = 0;
+                        bytesConsumed = 0;
                         return false;
                     }
-                    else
+                    int candidate = value * 10 + (nextByteVal); // parse the current digit to an int and add it to the left-shifted value
+                    value = candidate;
+                    bytesConsumed++; // increment the number of bytes consumed, then loop
+                }
+
+                if (negative) // We check if the value is negative at the very end to save on comp time
+                {
+                    value = -value;
+                }
+                return true;
+            }
+            else if (cultureAndEncodingInfo.IsInvariantUtf16)
+            {
+                if (utf8Text[index] == '-')
+                {
+                    negative = true;
+                    signed = true;
+                    index += 2;
+                    bytesConsumed += 2;
+                }
+                else if (utf8Text[index] == '+')
+                {
+                    signed = true;
+                    index += 2;
+                    bytesConsumed += 2;
+                }
+
+                for (int byteIndex = index; byteIndex < utf8Text.Length - 1; byteIndex += 2) // loop through the byte array two bytes at a time for UTF-16
+                {
+                    byte nextByte = utf8Text[byteIndex];
+                    byte byteAfterNext = utf8Text[byteIndex + 1];
+                    byte nextByteVal = (byte)(utf8Text[byteIndex] - '0');
+                    if (nextByteVal > 9 || byteAfterNext != 0) // if the second byte isn't zero, this isn't an ASCII-equivalent code unit and we can quit here
                     {
-                        if (negative && value != Int32.MinValue) // We check if the value is negative at the very end to save on comp time
+                        if (bytesConsumed == 2 && signed) // if the first character happened to be a '-' or a '+', we reset the byte counter so logic proceeds as normal
+                        {
+                            bytesConsumed = 0;
+                        }
+                        if (bytesConsumed == 0) // check to see if we've processed any digits at all
+                        {
+                            value = default(int); // if we haven't, set value to 0 and return false
+                            return false;
+                        }
+                        else
                         {
-                            value = -value;
+                            if (negative && value != Int32.MinValue) // We check if the value is negative at the very end to save on comp time
```

Comment:
you are right. we can simplify the implementation by just parsing the number without checking at all for the sign and after we finish parsing we can set the sign and check. this is like the implementation in http://index/#mscorlib/system/number.cs,552